htlcswitch/quiescer.go

```diff
@@ -0,0 +1,365 @@
+	// is important because the quiescer is responsible for constructing the
+	// messages we send out and the ChannelID is a key field in that
+	// message.
+	chanID lnwire.ChannelID
+
+	// channelInitiator indicates which ChannelParty originally opened the
+	// channel. This is used to break ties when both sides of the channel
+	// send Stfu claiming to be the initiator.
+	channelInitiator lntypes.ChannelParty
+
+	// numPendingUpdates is a function that returns the number of pending
+	// originated by the party in the first argument that have yet to be
+	// committed to the commitment transaction held by the party in the
+	// second argument.
+	numPendingUpdates func(lntypes.ChannelParty,
+		lntypes.ChannelParty) uint64
+
+	// sendMsg is a function that can be used to send an Stfu message over
+	// the wire.
+	sendMsg func(lnwire.Stfu) error
+}
+
+// quiescerLive is a state machine that tracks progression through the
+// quiescence protocol.
+type quiescerLive struct {
+	cfg quiescerCfg
+
+	// localInit indicates whether our path through this state machine was
+	// initiated by our node. This can be true or false independently of
+	// remoteInit.
+	localInit bool
+
+	// remoteInit indicates whether we received Stfu from our peer where the
+	// message indicated that the remote node believes it was the initiator.
+	// This can be true or false independently of localInit.
+	remoteInit bool
+
+	// sent tracks whether or not we have emitted Stfu for sending.
+	sent bool
+
+	// received tracks whether or not we have received Stfu from our peer.
+	received bool
+
+	// activeQuiescenceRequest is a possibly None Request that we should
+	// resolve when we complete quiescence.
+	activeQuiescenceReq fn.Option[StfuReq]
+
+	// resumeQueue is a slice of hooks that will be called when the quiescer
+	// is resumed. These are actions that needed to be deferred while the
+	// channel was quiescent.
+	resumeQueue []func()
+}
+
+// newQuiescer creates a new quiescer for the given channel.
+func newQuiescer(cfg quiescerCfg) quiescer {
+	return &quiescerLive{
+		cfg: cfg,
+	}
+}
+
+// recvStfu is called when we receive an Stfu message from the remote.
+func (q *quiescerLive) recvStfu(msg lnwire.Stfu) error {
+	// At the time of this writing, this check that we have already received
+	// an Stfu is not strictly necessary, according to the specification.
+	// However, it is fishy if we do and it is unclear how we should handle
+	// such a case so we will err on the side of caution.
+	if q.received {
+		return fmt.Errorf("%w for channel %v", ErrStfuAlreadyRcvd,
+			q.cfg.chanID)
+	}
+
+	if !q.canRecvStfu() {
+		return fmt.Errorf("%w for channel %v", ErrPendingRemoteUpdates,
+			q.cfg.chanID)
+	}
+
+	q.received = true
+
+	// If the remote party sets the initiator bit to true then we will
+	// remember that they are making a claim to the initiator role. This
+	// does not necessarily mean they will get it, though.
+	q.remoteInit = msg.Initiator
```

Comment:
looks like we are missing this check from the specs,
> if it is replying to an stfu:
MUST set initiator to 0
otherwise:
MUST set initiator to 1

Reply:
There's no way to ensure this. They could have sent theirs at the same time.